pwndbg/aglib/file.py

```diff
@@ -0,0 +1,178 @@
+    """
+    Downloads the specified file from the system where the current process is
+    being debugged.
+
+    If the `path` is prefixed with "target:" the prefix is stripped
+    (to support remote target paths properly).
+
+    If the `try_local_path` is set to `True` and the `path` exists locally and "target:" prefix is not present, it will return the local path instead of downloading the file.
+
+    Returns:
+        The local path to the file
+    """
+    assert path.startswith(("/", "./", "../")) or path.startswith(
+        "target:"
+    ), "get_file called with incorrect path"
+
+    has_target_prefix = path.startswith("target:")
+    if has_target_prefix:
+        path = path[7:]  # len('target:') == 7
+
+    local_path = path
+    qemu_root = pwndbg.aglib.qemu.root()
+
+    if qemu_root:
+        return os.path.join(qemu_root, path)
+
+    elif pwndbg.aglib.remote.is_remote():
+        if not pwndbg.aglib.qemu.is_qemu():
+            if try_local_path and not has_target_prefix and os.path.exists(local_path):
+                return local_path
+            local_path = tempfile.mktemp(dir=remote_files_dir())
+            try:
+                pwndbg.dbg.selected_inferior().download_remote_file(path, local_path)
+            except pwndbg.dbg_mod.Error as e:
+                # This module originally raised this as an OSError.
+                raise OSError(e)
+        else:
+            print(
+                message.warn(
+                    f"pwndbg.aglib.file.get_file({path}) returns local path as we can't download file from QEMU"
+                )
+            )
+
+    return local_path
+
+
+def get(path: str) -> bytes:
+    """
+    Retrieves the contents of the specified file on the system
+    where the current process is being debugged.
+
+    Returns:
+        A byte array, or None.
+    """
+    local_path = get_file(path)
+
+    try:
+        with open(local_path, "rb") as f:
+            return f.read()
+    except Exception:
+        return b""
+
+
+def readlink(path):
+    """readlink(path) -> str
+
+    Read the link specified by 'path' on the system being debugged.
+
+    Handles local, qemu-usermode, and remote debugging cases.
+    """
+    is_qemu = pwndbg.aglib.qemu.is_qemu_usermode()
+
+    if is_qemu:
+        if not os.path.exists(path):
+            # The or "" is needed since .root() may return None
+            # Then we just use the path (it can also be absolute too)
+            path = os.path.join(pwndbg.aglib.qemu.root() or "", path)
+
+    if is_qemu or not pwndbg.aglib.remote.is_remote():
+        try:
+            return os.readlink(path)
+        except Exception:
+            return ""
+
+    #
+    # Hurray unexposed packets!
+    #
+    # The 'vFile:readlink:' packet does exactly what it sounds like,
+    # but there is no API exposed to do this and there is also no
+    # command exposed... so we have to send the packet manually.
+    #
+    cmd = "vFile:readlink:%s"
```

Comment:
this code should go to dbg/gdb.py probably

Reply:
LLDB can handle this path gracefully. It doesn't seem to implement that packet, but it fails in a way we can handle. If it ever implements it, we get support for it almost for free (i.e. we have to change a few of the strings we're expecting in the following code, but it should work).